File: src/Algorithms/Prime Number/Sieve of Eratosthenes/code.js
// import visualization libraries {
const {
  Tracer,
  Array1DTracer,
  LogTracer,
  Layout,
  VerticalLayout,
} = require("algorithm-visualizer");
// }

const N = 30;
const a = [];
const b = [];
for (let i = 1; i <= N; i++) {
  a.push(i);
  b.push(0);
}

// define tracer variables {
const tracer = new Array1DTracer("Sieve");
tracer.set(a);
const logger = new LogTracer();
Layout.setRoot(new VerticalLayout([tracer, logger]));
Tracer.delay();
// }

// visualize {
logger.println("1 is not prime");
tracer.patch(0);
Tracer.delay();
// }
for (let i = 2; i <= N; i++) {
  if (b[i] === 0) {
    // visualize {
    tracer.select(i - 1);
    Tracer.delay();
    tracer.selectTrue(i - 1);
    // }
    logger.println(`${i} is not marked, so it is prime`);
    for (let j = i + i; j <= N; j += i) {
      b[j] = 1;
      // visualize {
      tracer.patch(j - 1);
      Tracer.delay();
      // }
      logger.println(`${j} is a multiple of ${i} so it is marked as red`);
    }
  }
}

logger.println(`The Green marked numbers are the prime numbers from 1 to ${N}`);

File: src/Algorithms/Prime Number/Sieve of Eratosthenes/code.py
from AlgorithmVisualizer import Tracer, Array1DTracer, LogTracer, Layout, VerticalLayout
n = 30
a = [i for i in range(1, n+1)]
b = [0 for _ in range(n+1)]

# defining variables {
tracer = Array1DTracer('Sieve')
tracer.set(a)
logger = LogTracer()
Layout.setRoot(VerticalLayout([tracer, logger]))
Tracer.delay()
# }

logger.println("1 is not prime So it marked red")
# visualize {
tracer.patch(0)
Tracer.delay()
# }

for i in range(2, n+1):
    if (b[i] == 0):
        tracer.select(i - 1)
        Tracer.delay()
        tracer.selectTrue(i - 1)
        logger.println(str(i)+' is not marked, so it is prime')
        for j in range(i+i,n+1,i):
            b[j]=1
            # visualize {
            tracer.patch(j - 1)
            Tracer.delay()
            # }
            logger.println(str(j)+' is a multiple of '+str(i)+' so it is marked as red')
logger.println(
    'The Green marked numbers are the prime numbers from 1 to '+str(n))
